### ISSUE TITLE
[SDL] PS4 Buttons Not Mapped or Doubly Mapped

### ISSUE THREAD

**Issue by AxiomVerge** [flibitijibibo/FNA#293]
When using a PS4 controller,there are these discrepancies:
-The right shoulder button appears not to work
-Pressing R2 also presses R3
-Pressing L2 also presses L3
-The big middle button appears  not to be mapped (not sure where it should be; Monogame PS4 maps it to "Buttons.BigButton")

**Comment by flibitijibibo** [comment]
fnalibs.tar.bz2 should contain an SDL2.dll at the moment; this is because SDL 2.0.3 shipped with a bad DS4 configuration on Windows. The custom-built DLL should fix this.

BigButton is not bound at the moment, let me fix this real quick (never thought I'd find someone using it since it brings up the guide on 360/Steam!).

**Comment by flibitijibibo** [comment]
Here's BigButton:

https://github.com/flibitijibibo/FNA/commit/b1cf0de942f57a5ad96149fd0f4d0b5619c714bd

Let me know if there's still something missing.

**Comment by flibitijibibo** [comment]
Should probably post for reference, in case someone else finds this issue:

https://hg.libsdl.org/SDL/rev/05237b5e8366

The red line is the bad config, the green line is the correct config. Was committed _just_ after 2.0.3, sadly. :(

**Comment by AxiomVerge** [comment]
Okay, the version of SDL2.dll I'm using is the one from 3/15/2014 that was in fnalibs archive... but even the brand new one had this issue.  It is actually weirdly intermittent - sometimes the Cross button is also mapped to L1 (so when you press Cross it also registers L1 as pressed), but not always.

I haven't installed the DS4 Tool software; am I supposed to?

Not sure if this makes a difference but I'm using the DS4 that comes with the devkit; it looks just like a retail DS4 but it might have some difference.

**Comment by flibitijibibo** [comment]
If we're getting any input at all then it's probably getting recognized by SDL2; the default fallback config deliberately starts with null bindings.

SDL2 Win32 uses a hybrid of XInput/DInput depending on what hardware's connected, so in this case it should be reading directly from the driver without any additional middlewares.

My guess is the intermittent behavior is due to the way the 2.0.3 mapping double-bound the buttons, so in theory it could be either one... but why the custom-built binary isn't working, I'm not sure.

@elisee, do you have your SDL2.dll hosted somewhere? Yours is a lot newer than mine, and it might take a while for me to make a new build...

**Comment by flibitijibibo** [comment]
Actually, scratch that, we've got Hg revisions hosted on the buildbot:

https://buildbot.libsdl.org/sdl-builds/sdl-visualstudio/

The latest zip should have a new SDL2.dll in the lib/win32/ folder. If this doesn't work, I'd check what Steam Big Picture configures your controller to. If you've configured this already, the string will be in Steam/config/config.vdf, under the "SDL_GamepadBind" entry.

**Comment by AxiomVerge** [comment]
Well, I'm not sure what's going on, because I grabbed the dll for this folder, and still had the double mapped buttons issue.  So I rebooted, thinking, I dunno, Visual Studio was holding onto the old dll or something (thinking it could be related to that case when VS refuses to build because it's still open in another application, which turns out to be Visual Studio itself), but that didn't work.

So I opened up big picture, and the PS4 controller seemed to work fine.  However, there wasn't an SDL_GamepadBind entry anywhere in my steam directory (I did a "find in files" to verify).  So for kicks I copied the SDL2.dll in the Steam folder into my output folder; it was still broken.  Then I remembered it Shovel Knight was supposed to be compatible with DS4, so, I opened that and started remapping; all buttons worked fine, right shoulder, left shoulder, analog press, etc.  So I tried copying the SDL2.dll from ShovelKnight to AV and ... still no worky!

BUT ... Resigned to failure, I eventually went back to the buildbot revision above, just because if I was going to have a broken dll, I wanted the *latest* broken dll ... 

... and NOW IT WORKS.

I'm rather baffled by this.  I'm sort of expecting it to randomly be broken again later on, just because I have no explanation for how it could be fixed after going through such effort.  But, for now, it is working :+1:

**Comment by flibitijibibo** [comment]
Yeah, I bet everyone's still on 2.0.3... that binding was a really awful oversight, and 2.0.4's delays haven't helped. It's been messing with Windows titles forever, with some people just using SDL_GameControllerAddMapping to override it (Transistor's GamePad implementation still does this, I think).

Once 2.0.4's out we _should_ be able to put all this mess behind us, but for now it looks like the Hg buildbot will be our best bet.

If this comes up again feel free to poke this thread again.

**Comment by AxiomVerge** [comment]
Well, just thought I'd mention - it came back again, but, I unplugged the usb cable from the top of the DS4 and back again, and it was fixed.  I think my USB cable or the socket on the controller might be bad; it doesn't feel secure.  So, I am now wondering if the bug would have happened even with the 2.0.3 SDL if I just used a different controller (I only have one but am going to switch cables now).
